tests/unit/s2n_tls12_encrypted_extensions_test.c

```diff
@@ -0,0 +1,141 @@
+            EXPECT_SUCCESS(s2n_tls12_encrypted_extensions_send(client_conn));
+
+        }
+
+        /* Sends the npn extension */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_allow_all_response_extensions(client_conn));
+            client_conn->actual_protocol_version = S2N_TLS12;
+
+            struct s2n_stuffer *stuffer = &client_conn->handshake.io;
+
+            char protocol[] = { "http/1.1" };
+            uint8_t protocol_len = strlen(protocol);
+            EXPECT_MEMCPY_SUCCESS(client_conn->application_protocol, protocol, protocol_len + 1);
+
+            EXPECT_SUCCESS(s2n_tls12_encrypted_extensions_send(client_conn));
+
+            uint16_t extension_list_size;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(stuffer, &extension_list_size));
+            EXPECT_NOT_EQUAL(extension_list_size, 0);
+            EXPECT_EQUAL(s2n_stuffer_data_available(stuffer), extension_list_size);
+
+            uint16_t extension_type;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint16(stuffer, &extension_type));
+            EXPECT_EQUAL(extension_type, s2n_npn_encrypted_extension.iana_value);
+        }
+    }
+
+    /* Test s2n_tls12_encrypted_extensions_recv */
+    {
+        /* Safety checks */
+        EXPECT_FAILURE(s2n_tls12_encrypted_extensions_recv(NULL));
+
+        /* Should fail for >=TLS1.3 */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+
+            /* Fails for TLS1.3 */
+            server_conn->actual_protocol_version = S2N_TLS13;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_tls12_encrypted_extensions_recv(server_conn), S2N_ERR_BAD_MESSAGE);
+
+            /* Succeeds for TLS1.2 */
+            server_conn->actual_protocol_version = S2N_TLS12;
+            EXPECT_SUCCESS(s2n_tls12_encrypted_extensions_recv(server_conn));
+        }
+
+        /* Should parse an empty list */
+        {
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            server_conn->actual_protocol_version = S2N_TLS12;
+
+            struct s2n_stuffer *stuffer = &server_conn->handshake.io;
+
+            /* Parse no data */
+            EXPECT_SUCCESS(s2n_tls12_encrypted_extensions_recv(server_conn));
+            EXPECT_EQUAL(s2n_stuffer_data_available(stuffer), 0);
+
+            /* Parse explicitly empty list */
+            EXPECT_SUCCESS(s2n_extension_list_send(S2N_EXTENSION_LIST_EMPTY, server_conn, stuffer));
+            EXPECT_SUCCESS(s2n_tls12_encrypted_extensions_recv(server_conn));
+            EXPECT_EQUAL(s2n_stuffer_data_available(stuffer), 0);
+        }
+
+        /* Should parse s2n_tls12_encrypted_extensions_send */
```

Comment:
Are there any malformed / failure cases we should check?

Reply:
Fuzz tests can catch unexpected issues, but we shouldn't treat them as a replacement for unit tests. Let test at least one "malformed" case.